src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,181 @@
+                Name: nameof(ToString),
+                IsOverride: true
+            };
+        }
+
+        private static bool IsDebuggerDisplayMethod(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol is
+            {
+                Arity: 0,
+                Parameters: { IsEmpty: true },
+                Name: DebuggerDisplayMethodName
+            };
+        }
+
+        private static bool IsClassOrStruct(ITypeSymbol typeSymbol)
+        {
+            switch (typeSymbol.TypeKind)
+            {
+                case TypeKind.Class:
+                case TypeKind.Struct:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private static bool HasDebuggerDisplayAttribute(ITypeSymbol typeSymbol, Compilation compilation)
+        {
+            return typeSymbol.GetAttributes()
+                .Select(data => data.AttributeClass)
+                .Contains(compilation.GetTypeByMetadataName("System.Diagnostics.DebuggerDisplayAttribute"));
+        }
+
+        private async Task<Document> ApplyAsync(Document document, TTypeDeclarationSyntax type, CancellationToken cancellationToken)
+        {
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var symbolicModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var newAttribute = generator
+                .Attribute("System.Diagnostics.DebuggerDisplayAttribute", symbolicModel.Compilation, generator.LiteralExpression("{" + DebuggerDisplayMethodName + "(),nq}"))
+                .WithAdditionalAnnotations(Simplifier.Annotation);
+
+            var modifiedType = generator.AddAttributes(type, newAttribute);
+
+            var typeSymbol = (ITypeSymbol)symbolicModel.GetDeclaredSymbol(type, cancellationToken);
+
+            if (!typeSymbol.GetMembers().OfType<IMethodSymbol>().Any(IsDebuggerDisplayMethod))
+            {
+                modifiedType = generator.AddMembers(modifiedType,
+                    generator.MethodDeclaration(
+                        DebuggerDisplayMethodName,
+                        returnType: generator.TypeExpression(SpecialType.System_String),
+                        accessibility: Accessibility.Private,
+                        statements: new[]
+                        {
+                            generator.ReturnStatement(generator.InvocationExpression(
+                                generator.MemberAccessExpression(generator.ThisExpression(), generator.IdentifierName("ToString"))))
+                        }));
+            }
+
+            modifiedType = modifiedType.WithAdditionalAnnotations(s_trackingAnnotation);
+
+            syntaxRoot = syntaxRoot.ReplaceNode(type, modifiedType);
+
+            syntaxRoot = await EnsureNamespaceImportAsync(
```

Comment:
you shouldn't need this with @YairHalberstadt 's recent work.  right @YairHalberstadt ?

Reply:
Yair's PR that still hasn't gotten a response in a while?

Reply:
It shouldn't be necessary once it's merged. It doesn't seem to be moving very fast but at the same time, I don't think it makes sense to partially implement a subset of the functionality.

Reply:
Well, I do need this for the time being so that the tests pass. :)